### PR TITLE
First pass at tasks to clean up abandoned EC2 resources

### PR DESCRIPTION
**Pull request by jmanero-r7**

### CODE REVIEW

File: lib/builderator/control/asg.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        end
+
+        def active_launch_configs(update = false)
+          return @active_launch_configs unless @active_launch_configs.nil? || update
+
+          @active_launch_configs = {}.tap do |configs|
+            launch_configs(true) if update
+            scaling_groups(update).each do |_, asg|
+              configs[asg.launch_configuration_name] = launch_configs[asg.launch_configuration_name]
+            end
```
Comment (athompson-r7) [inline lib/builderator/control/asg.rb]: What makes these active?  Just that they are referenced in an asg?

File: lib/builderator/control/ec2.rb [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+          return @unused_snapshots unless @unused_snapshots.nil? || update
+
+          @unused_snapshots = {}.tap do |unused|
+            (snapshots(options, update).keys - active_snapshots(options, update).keys).each do |s|
+              unused[s] = snapshots[s]
+            end
+          end
+        end
+
+        def active_volumes(_ = {}, update = false)
```
Comment (athompson-r7) [inline lib/builderator/control/ec2.rb]: Is `_` some sort of magic, because it's not used anywhere.

File: lib/builderator/tasks/clean.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      def configs
+        Control::ASG.unused_launch_configs.each do |l, _|
+          say_status :remove, "Launch Configuration #{ l }", :red
+        end
+      end
+
+      desc 'images', 'Deregister unused images'
+      option 'group-by',
+             :type => :array,
+             :desc => 'Tags/properties to group images by for pruning'
```
Comment (athompson-r7) [inline lib/builderator/tasks/clean.rb:42]: Is this unbounded, e.g. is the input well know based on tags, etc.?

File: lib/builderator/tasks/clean.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      option 'sort-by',
+             :type => :string,
+             :default => 'creation_date',
+             :desc => 'Tag/property to sort grouped images on'
+      option :keep,
+             :type => :numeric,
+             :default => 5,
+             :desc => 'Number of images in each group to keep'
```
Comment (athompson-r7) [inline lib/builderator/tasks/clean.rb:50]: Unless I am missing it, I'm not seeing anything that takes the number of images to keep into account.

Comment (jmanero-r7) [inline lib/builderator/tasks/clean.rb:50]: Yeah, I re-worked the underlying module structure to make this behavior more obvious. The logic lives in https://github.com/rapid7/builderator/blob/ec2-clean-tasks/lib/builderator/model/images.rb#L47 now.

File: lib/builderator/tasks/clean.rb (hunk 3 above)
Comment (jmanero-r7) [inline lib/builderator/tasks/clean.rb:42]: This is a good point. I'm thinking that we need to add a default limit to the number of resources that would be deleted. If the `unused` count exceeds that, the task will fail, requiring a human to verify the set before continuing.

File: lib/builderator/model.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        Util.filter(resources, filters)
+      end
+
+      def select(set = [])
+        resources.select { |k, _| set.include?(k) }
+      end
+
+      def reject(set)
+        resources.reject { |k, _| set.include?(k) }
+      end
```
Comment (athompson-r7) [inline lib/builderator/model.rb:28]: This will throw an exception if `set` is a `NilClass`.

File: lib/builderator/model/instances.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+require_relative '../util'
+
+module Builderator
+  module Model
+
+    def self.instances
+      @instances ||= Instances.new
+    end
+    ##
+    # Manage AusoScaling resources
```
Comment (athompson-r7) [inline lib/builderator/model/instances.rb]: Spelling...

**Comment by athompson-r7** [comment]
I think most of the class comment headers are lying, can you double check them?

File: lib/builderator/model/launch_configs.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        end
+      end
+
+      def images
+        resources.values.map { |l| l[:image] }
+      end
+
+      def in_use(_)
+        select(Model.scaling_groups.launch_configs)
+      end
```
Comment (athompson-r7) [inline lib/builderator/model/launch_configs.rb:43]: `_` method parameter is not used and should be removed.

**Comment by athompson-r7** [comment]
It looks as though the `fetch` instance method is never called, so the resources `attr_reader` would not be loaded.

**Comment by jmanero-r7** [comment]
`fetch` is aliased to `initialize` in resources' super calss by default

File: lib/builderator/model/launch_configs.rb (hunk 7 above)
Comment (jmanero-r7) [inline lib/builderator/model/launch_configs.rb:43]: This is an implementation of the parent's method. The interface can accept an `options` Hash, but this resource doesn't use it.

File: lib/builderator/tasks/clean.rb (hunk 3 above)
Comment (jmanero-r7) [inline lib/builderator/tasks/clean.rb:42]: Fixed in 072dd13

File: lib/builderator/model.rb (hunk 5 above)
Comment (jmanero-r7) [inline lib/builderator/model.rb:28]: I'm not too concerned about that TBH. If something fails to pass a filter-set, I'd rather an exception be thrown than the reject method fail to filter correctly.

**Pull request merged**